File: backend/src/main/java/com/NBE3_4_2_Team4/global/security/oauth2/tokenService/impl/KakaoTokenService.java
package com.NBE3_4_2_Team4.global.security.oauth2.tokenService.impl;

import com.NBE3_4_2_Team4.global.security.oauth2.token.service.OAuth2TokenService;
import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.ObjectMapper;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.http.*;
import org.springframework.stereotype.Component;
import org.springframework.util.LinkedMultiValueMap;
import org.springframework.util.MultiValueMap;
import org.springframework.web.client.RestClientException;
import org.springframework.web.client.RestTemplate;

@Slf4j
@Component
@RequiredArgsConstructor
public class KakaoTokenService implements OAuth2TokenService {
    @Value("${spring.security.oauth2.client.registration.kakao.clientId}")
    private String clientId;
    @Value("${spring.security.oauth2.client.provider.kakao.token-uri}")
    private String kakaoTokenUrl;

    private final RestTemplate restTemplate;
    private final ObjectMapper objectMapper = new ObjectMapper();

    public String getFreshAccessToken(String refreshToken) {
        HttpHeaders headers = new HttpHeaders();
        headers.setContentType(MediaType.APPLICATION_FORM_URLENCODED);

        MultiValueMap<String, String> body = new LinkedMultiValueMap<>();
        body.add("grant_type", "refresh_token");
        body.add("client_id", clientId);
        body.add("refresh_token", refreshToken);

        HttpEntity<MultiValueMap<String, String>> requestEntity = new HttpEntity<>(body, headers);
        try {
            ResponseEntity<String> response = restTemplate.postForEntity(kakaoTokenUrl, requestEntity, String.class);
            return objectMapper.readTree(response.getBody())
                    .get("access_token").asText();
        }catch (RestClientException | JsonProcessingException e) {
            log.error("error occurred while get accessToken for kakao. msg : {}",e.getMessage());
            return null;
        }
    }
}

File: backend/src/main/java/com/NBE3_4_2_Team4/global/security/oauth2/tokenService/impl/KakaoTokenService.kt
package com.NBE3_4_2_Team4.global.security.oauth2.tokenService.impl

import com.NBE3_4_2_Team4.global.security.oauth2.tokenService.OAuth2TokenService
import com.fasterxml.jackson.core.JsonProcessingException
import com.fasterxml.jackson.databind.ObjectMapper
import org.slf4j.LoggerFactory
import org.springframework.beans.factory.annotation.Value
import org.springframework.http.HttpEntity
import org.springframework.http.HttpHeaders
import org.springframework.http.MediaType
import org.springframework.http.ResponseEntity
import org.springframework.stereotype.Service
import org.springframework.util.LinkedMultiValueMap
import org.springframework.util.MultiValueMap
import org.springframework.web.client.HttpClientErrorException
import org.springframework.web.client.RestClientException
import org.springframework.web.client.RestTemplate

@Service
class KakaoTokenService(
    private val restTemplate: RestTemplate,
    private val objectMapper: ObjectMapper
): OAuth2TokenService {
    private val log = LoggerFactory.getLogger(KakaoTokenService::class.java)
    @Value("\${spring.security.oauth2.client.registration.kakao.clientId}")
    lateinit var clientId: String

    @Value("\${spring.security.oauth2.client.provider.kakao.token-uri}")
    lateinit var kakaoTokenUri: String

    override fun getFreshAccessToken(refreshToken: String?): String? {
        val headers = HttpHeaders()
        headers.contentType = MediaType.APPLICATION_FORM_URLENCODED

        val body: MultiValueMap<String, String> = LinkedMultiValueMap()
        body.add("grant_type", "refresh_token")
        body.add("client_id", clientId)
        body.add("refresh_token", refreshToken)

        val requestEntity = HttpEntity(body, headers)
        try {
            val response: ResponseEntity<String> = restTemplate.postForEntity(
                kakaoTokenUri, requestEntity,
                String::class.java
            )
            return objectMapper.readTree(response.body)["access_token"].asText()
        }  catch (e: Exception) {
            log.error("error occurred while getting accessToken for google. msg: {}", e.message)
            return null
        }
    }
}